editoast/src/views/infra/mod.rs

```diff
@@ -141,6 +146,51 @@ async fn create(
     .await
 }
 
+/// Duplicate an infra
+#[post("/<infra>/clone")]
+async fn clone(infra: i64, conn: DBConnection) -> ApiResult<Custom<Json<i64>>> {
+    let mut model_table_names = vec![];
+    let mut layer_table_names = vec![];
+    for object in ObjectType::iter() {
+        model_table_names.push(object.get_table().to_string());
+        if object != ObjectType::SwitchType && object != ObjectType::Route {
+            layer_table_names.push(object.get_geometry_layer_table().unwrap().to_string());
+        }
+    }
+    let cloned_infra = conn
+        .run::<_, ApiResult<_>>(move |conn| {
+            let cloned_infra = Infra::clone(infra, conn)?;
+            Ok(cloned_infra)
+        })
+        .await
+        .unwrap();
+    let loop_model = async {
+        for table in model_table_names {
+            conn.run(move |conn| {
+            sql_query(format!(
+                "INSERT INTO {table}(id, obj_id,data,infra_id) SELECT nextval('{table}_id_seq'), obj_id,data,$1 FROM {table} WHERE infra_id=$2"
+            ))
+            .bind::<BigInt, _>(cloned_infra.id)
+            .bind::<BigInt, _>(infra)
+            .execute(conn)})
+            .await.ok();
+        }
+    };
+
+    let loop_layer = async {
+        for table in layer_table_names {
+            conn.run(move |conn| {sql_query(format!(
+                "INSERT INTO {table}(id, obj_id,geographic,schematic,infra_id) SELECT nextval('{table}_id_seq'), obj_id,geographic,schematic,$1 FROM {table} WHERE infra_id=$2"
+            ))
+            .bind::<BigInt, _>(cloned_infra.id)
+            .bind::<BigInt, _>(infra)
+            .execute(conn)})
+            .await.ok();
+        }
+    };
+    join!(loop_model, loop_layer);
```

Comment:
Here, what you do is run two sets of queries simultaneously. One for the object data and another for the generated data. The problem is that the queries are executed one after the other within a set.

----

The expected behavior is to [join_all](https://docs.rs/futures/latest/futures/future/fn.join_all.html) the futures (meaning the object duplication) at once. So we use all the possible db connection as possible.